cpp/include/cudf/table/experimental/row_operators.cuh

```diff
@@ -0,0 +1,426 @@
+#include <cudf/utilities/traits.hpp>
+#include <cudf/utilities/type_dispatcher.hpp>
+
+#include <thrust/equal.h>
+#include <thrust/swap.h>
+#include <thrust/transform_reduce.h>
+
+#include <cuda/std/tuple>
+#include <cuda/std/utility>
+
+#include <limits>
+#include <memory>
+#include <utility>
+
+namespace cudf {
+namespace experimental {
+
+/**
+ * @brief A map from cudf::type_id to cudf type that excludes LIST and STRUCT types.
+ *
+ * To be used with type_dispatcher in place of the default map, when it is required that STRUCT and
+ * LIST map to void.
+ *
+ * Usage:
+ * @code
+ * type_dispatcher<non_nested_id_to_type>(data_type(), functor{});
+ * @endcode
+ */
+template <cudf::type_id t>
+struct non_nested_id_to_type {
+  using type = std::conditional_t<cudf::is_nested(data_type(t)), void, id_to_type<t>>;
+};
+
+namespace lex {
+
+/**
+ * @brief Performs a relational comparison between two elements in two columns.
+ *
+ * @tparam Nullate A cudf::nullate type describing how to check for nulls.
+ */
+template <typename Nullate>
+class element_comparator {
+ public:
+  /**
+   * @brief Construct type-dispatched function object for performing a
+   * relational comparison between two elements.
+   *
+   * @note `lhs` and `rhs` may be the same.
+   *
+   * @param has_nulls Indicates if either input column contains nulls.
+   * @param lhs The column containing the first element
+   * @param rhs The column containing the second element (may be the same as lhs)
+   * @param null_precedence Indicates how null values are ordered with other values
+   * @param depth The depth of the column if part of a nested column @see preprocessed_table::depths
+   */
+  __host__ __device__ element_comparator(Nullate has_nulls,
+                                         column_device_view lhs,
+                                         column_device_view rhs,
+                                         null_order null_precedence,
+                                         int depth = 0)
+    : _lhs{lhs}, _rhs{rhs}, _nulls{has_nulls}, _null_precedence{null_precedence}, _depth{depth}
+  {
+  }
+
+  __host__ __device__ element_comparator(Nullate has_nulls,
+                                         column_device_view lhs,
+                                         column_device_view rhs)
+    : _lhs{lhs}, _rhs{rhs}, _nulls{has_nulls}
+  {
+  }
+
+  /**
+   * @brief Performs a relational comparison between the specified elements
+   *
+   * @param lhs_element_index The index of the first element
+   * @param rhs_element_index The index of the second element
+   * @return Indicates the relationship between the elements in the `lhs` and `rhs` columns, along
+   * with the depth at which a null value was encountered.
+   */
+  template <typename Element, CUDF_ENABLE_IF(cudf::is_relationally_comparable<Element, Element>())>
+  __device__ cuda::std::pair<weak_ordering, int> operator()(
+    size_type lhs_element_index, size_type rhs_element_index) const noexcept
```

Comment:
What's the purpose of cv qualifying pod args?

Reply:
I have on occasion observed that in more complex kernels (specifically in the context of ASTs and/or joins) that const qualifying POD arguments helps the compiler when it's already overloaded with too much complexity. Sometimes the optimizer seems to fail to recognize that it can avoid making copies leading to extra register pressure etc. But that's very anecdotal and probably not important 99% of the time.